modules/security/pam.nix

```diff
@@ -0,0 +1,62 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.security.pam;
+
+  # Implementation Notes
+  #
+  # We don't use `environment.etc` because this would require that the user manually delete
+  # `/etc/pam.d/sudo` which seems unwise given that applying the nix-darwin configuration requires
+  # sudo. We also can't use `system.patchs` since it only runs once, and so won't patch in the
+  # changes again after OS updates (which remove modifications to this file).
+  #
+  # As such, we resort to line addition/deletion in place using `sed`. We add a comment to the
+  # added line that includes the name of the option, to make it easier to identify the line that
+  # should be deleted when the option is disabled.
+  mkSudoTouchIdAuthScript = isEnabled:
+  let
+    file   = "/etc/pam.d/sudo";
+    option = "security.pam.enableSudoTouchIdAuth";
+    sed = "${pkgs.gnused}/bin/sed";
+  in ''
+    ${if isEnabled then ''
+      # Enable sudo Touch ID authentication, if not already enabled
+      if ! grep 'pam_tid.so' ${file} > /dev/null; then
+        ${sed} -i '2i\
+      auth       sufficient     pam_tid.so # nix-darwin: ${option}
+        ' ${file}
+      fi
+    '' else ''
+      # Disable sudo Touch ID authentication, if added by nix-darwin
+      if grep '${option}' ${file} > /dev/null; then
+        ${sed} -i '/${option}/d' ${file}
+      fi
+    ''}
+  '';
+in
+
+{
+  options = {
+    security.pam.enableSudoTouchIdAuth = mkEnableOption ''
+      Enable sudo authentication with Touch ID
+
+      When enabled, this option adds the following line to /etc/pam.d/sudo:
+
+          auth       sufficient     pam_tid.so
+
+      (Note that macOS resets this file when doing a system update. As such, sudo
+      authentication with Touch ID won't work after a system update until the nix-darwin
+      configuration is reapplied.)
```

Comment:
Apologies for bumping this three year old PR, but I figured this is the best place to ask this.

Isn't this not actually true due to the system activation service [being enabled by default](https://github.com/LnL7/nix-darwin/blob/eb2b9b64238349bd351561e32e260cac15db6f9a/modules/services/activate-system/default.nix#L13), which will reapply the PAM edits upon boot?

Assuming I'm correct, I think that leaving this untrue (by default) clarification in the documentation will just confuse users.

Reply:
What you linked doesn't run the entire activation script, it just runs part of it

https://github.com/LnL7/nix-darwin/blob/eb2b9b64238349bd351561e32e260cac15db6f9a/modules/services/activate-system/default.nix#L37-L39

I suppose this could have been updated to include the pam activation script too if someone thought about it. I actually didn't even realize this was here. Though at this point it's worth noting that https://github.com/LnL7/nix-darwin/pull/787 changes how all of this works.